Debian Jessie apt-get update fails since Debian packages have been moved

https://twitter.com/debian/status/1109080168318926851
https://lists.debian.org/debian-devel-announce/2019/03/msg00006.html

Debian as archived Jessie packages, so I'm not able to run Jepsen tests via the Docker setup anymore.

I'm trying to run Jepsen tests with the Docker setup using the `tutum/debian:jessie` image (the default):
 
```
./up.sh --dev
```

Tests no longer start since Debian has archived the Jessie packages. E.g., I try to run this test:

```
lein run test --local-binary ./dgraph --force-download --upsert-schema --workload bank --nemesis none --time-limit 300
```

And I get this Jepsen error with `apt-get update`:

```
Caused by: clojure.lang.ExceptionInfo: throw+: {:type :jepsen.control/nonzero-exit, :cmd "sudo -S -u root bash -c \"cd /; apt-get update\"", :exit 100, :out "Hit http://security.debian.org jessie/updates InRelease\nIgn http://httpredir.debian.org jessie InRelease\nIgn http://httpredir.debian.org jessie-updates InRelease\nHit http://httpredir.debian.org jessie Release.gpg\nIgn http://httpredir.debian.org jessie-updates Release.gpg\nHit http://httpredir.debian.org jessie Release\nIgn http://httpredir.debian.org jessie-updates Release\nErr http://httpredir.debian.org jessie-updates/main amd64 Packages\n  \nErr http://httpredir.debian.org jessie-updates/main amd64 Packages\n  \nErr http://httpredir.debian.org jessie-updates/main amd64 Packages\n  \nErr http://httpredir.debian.org jessie-updates/main amd64 Packages\n  \nErr http://httpredir.debian.org jessie-updates/main amd64 Packages\n  404  Not Found\nGet:1 http://security.debian.org jessie/updates/main amd64 Packages [822 kB]\nGet:2 http://httpredir.debian.org jessie/main amd64 Packages [9098 kB]\nFetched 9920 kB in 6s (1441 kB/s)\n", :err "W: Failed to fetch http://httpredir.debian.org/debian/dists/jessie-updates/main/binary-amd64/Packages  404  Not Found\n\nE: Some index files failed to download. They have been ignored, or old ones used instead.\n", :host "n1", :action {:cmd "sudo -S -u root bash -c \"cd /; apt-get update\"", :in "root\n"}}
	at slingshot.support$stack_trace.invoke(support.clj:201) ~[knossos-0.3.4.jar:na]
	at jepsen.control$throw_on_nonzero_exit.invokeStatic(control.clj:127) ~[jepsen-0.1.12.jar:na]
	at jepsen.control$throw_on_nonzero_exit.invoke(control.clj:122) ~[jepsen-0.1.12.jar:na]
	at jepsen.control$exec_STAR_.invokeStatic(control.clj:163) ~[jepsen-0.1.12.jar:na]
	at jepsen.control$exec_STAR_.doInvoke(control.clj:160) ~[jepsen-0.1.12.jar:na]
	at clojure.lang.RestFn.applyTo(RestFn.java:137) [clojure-1.9.0.jar:na]
	at clojure.core$apply.invokeStatic(core.clj:657) ~[clojure-1.9.0.jar:na]
	at clojure.core$apply.invoke(core.clj:652) ~[clojure-1.9.0.jar:na]
	at jepsen.control$exec.invokeStatic(control.clj:179) ~[jepsen-0.1.12.jar:na]
	at jepsen.control$exec.doInvoke(control.clj:173) ~[jepsen-0.1.12.jar:na]
	at clojure.lang.RestFn.invoke(RestFn.java:421) [clojure-1.9.0.jar:na]
	at jepsen.os.debian$update_BANG_.invokeStatic(debian.clj:36) ~[na:na]
	at jepsen.os.debian$update_BANG_.invoke(debian.clj:33) ~[na:na]
	at jepsen.os.debian$maybe_update_BANG_.invokeStatic(debian.clj:42) ~[na:na]
	at jepsen.os.debian$maybe_update_BANG_.invoke(debian.clj:38) ~[na:na]
	at jepsen.os.debian.Debian.setup_BANG_(debian.clj:145) ~[na:na]
	at jepsen.os$fn__2294$G__2288__2298.invoke(os.clj:4) ~[jepsen-0.1.12.jar:na]
	at jepsen.os$fn__2294$G__2287__2303.invoke(os.clj:4) ~[jepsen-0.1.12.jar:na]
	at clojure.core$partial$fn__5561.invoke(core.clj:2617) ~[clojure-1.9.0.jar:na]
	at jepsen.control$on_nodes$fn__2893.invoke(control.clj:384) ~[jepsen-0.1.12.jar:na]
	at clojure.lang.AFn.applyToHelper(AFn.java:154) ~[clojure-1.9.0.jar:na]
	at clojure.lang.AFn.applyTo(AFn.java:144) ~[clojure-1.9.0.jar:na]
	at clojure.core$apply.invokeStatic(core.clj:657) ~[clojure-1.9.0.jar:na]
	at clojure.core$with_bindings_STAR_.invokeStatic(core.clj:1965) ~[clojure-1.9.0.jar:na]
	at clojure.core$with_bindings_STAR_.doInvoke(core.clj:1965) ~[clojure-1.9.0.jar:na]
	at clojure.lang.RestFn.applyTo(RestFn.java:142) [clojure-1.9.0.jar:na]
	at clojure.core$apply.invokeStatic(core.clj:661) ~[clojure-1.9.0.jar:na]
	at clojure.core$bound_fn_STAR_$fn__5471.doInvoke(core.clj:1995) ~[clojure-1.9.0.jar:na]
	at clojure.lang.RestFn.invoke(RestFn.java:408) [clojure-1.9.0.jar:na]
	at jepsen.util$real_pmap$launcher__2044$fn__2045.invoke(util.clj:50) ~[jepsen-0.1.12.jar:na]
	at clojure.core$binding_conveyor_fn$fn__5476.invoke(core.clj:2022) ~[clojure-1.9.0.jar:na]
	at clojure.lang.AFn.call(AFn.java:18) ~[clojure-1.9.0.jar:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_191]
	at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_191]
```

And when I run `apt-get update` from a node directly, I get errors saying the debian repos aren't available:
```
$ docker exec -it jepsen-n1 bash     
root@n1:/# sudo apt-get update
Hit http://security.debian.org jessie/updates InRelease
Ign http://httpredir.debian.org jessie InRelease              
Ign http://httpredir.debian.org jessie-updates InRelease      
Get:1 http://security.debian.org jessie/updates/main amd64 Packages [822 kB]
Hit http://httpredir.debian.org jessie Release.gpg     
Ign http://httpredir.debian.org jessie-updates Release.gpg          
Hit http://httpredir.debian.org jessie Release                      
Ign http://httpredir.debian.org jessie-updates Release
Get:2 http://httpredir.debian.org jessie/main amd64 Packages [9098 kB]
Err http://httpredir.debian.org jessie-updates/main amd64 Packages
  
Err http://httpredir.debian.org jessie-updates/main amd64 Packages
  
Err http://httpredir.debian.org jessie-updates/main amd64 Packages
  
Err http://httpredir.debian.org jessie-updates/main amd64 Packages                                                                                                                          
  
Err http://httpredir.debian.org jessie-updates/main amd64 Packages                                                                                                                          
  404  Not Found
Fetched 9920 kB in 6s (1483 kB/s)                                                                                                                                                           
W: Failed to fetch http://httpredir.debian.org/debian/dists/jessie-updates/main/binary-amd64/Packages  404  Not Found

E: Some index files failed to download. They have been ignored, or old ones used instead.
```

I also tried `./up.sh --dev --ubuntu`, which I thought would work since I figured it'd use Ubuntu instead of Debian, but I get the same Debian errors in the worker nodes. It seems like `--ubuntu` still created Debian-based containers instead of Ubuntu-based ones.

--- a/docker/control/bashrc
+++ b/docker/control/bashrc
@@ -1,3 +1,6 @@
+eval $(ssh-agent) &> /dev/null
+ssh-add /root/.ssh/id_rsa &> /dev/null
+
 cat <<EOF
 Welcome to Jepsen on Docker
 ===========================

--- a/docker/node/Dockerfile
+++ b/docker/node/Dockerfile
@@ -12,8 +12,6 @@ RUN chmod +x /*.sh
 
 ENV AUTHORIZED_KEYS **None**
 
-RUN sed -i '/jessie-updates/d' /etc/apt/sources.list
-
 RUN apt-get update
 RUN apt install -y apt-transport-https
 RUN apt install -y software-properties-common

--- a/docker/node/Dockerfile
+++ b/docker/node/Dockerfile
@@ -1,4 +1,25 @@
-# FIXME: tutum/debian will be deprecated soon: https://github.com/tutumcloud/tutum-debian/blob/master/README.md
-FROM tutum/debian:jessie
+FROM debian:stretch
 
-RUN rm /etc/apt/apt.conf.d/docker-clean && apt-get update && apt-get install -y bzip2 curl faketime iproute iptables iputils-ping libzip2 logrotate man man-db net-tools ntpdate psmisc python rsyslog sudo sysvinit sysvinit-core sysvinit-utils tar unzip vim wget && apt-get remove -y --purge --auto-remove systemd
+# Install packages
+RUN apt-get update && \
+apt-get -y install openssh-server pwgen && \
+mkdir -p /var/run/sshd && \
+sed -i "s/UsePrivilegeSeparation.*/UsePrivilegeSeparation no/g" /etc/ssh/sshd_config && \
+sed -i "s/PermitRootLogin without-password/PermitRootLogin yes/g" /etc/ssh/sshd_config
+
+ADD run.sh /run.sh
+RUN chmod +x /*.sh
+
+ENV AUTHORIZED_KEYS **None**
+
+RUN sed -i '/jessie-updates/d' /etc/apt/sources.list
+
+RUN apt-get update
+RUN apt install -y apt-transport-https
+RUN apt install -y software-properties-common
+
+# Install Jepsen deps
+RUN apt-get install -y build-essential bzip2 curl faketime iproute iptables iputils-ping libzip4 logrotate man man-db net-tools ntpdate psmisc python rsyslog sudo tar unzip vim wget && apt-get remove -y --purge --auto-remove systemd
+
+EXPOSE 22
+CMD ["/run.sh"]

--- a/docker/node/Dockerfile
+++ b/docker/node/Dockerfile
@@ -1,3 +1,4 @@
+# Based on the deprecated `https://github.com/tutumcloud/tutum-debian`
 FROM debian:stretch
 
 # Install packages
@@ -11,6 +12,9 @@ ADD run.sh /run.sh
 RUN chmod +x /*.sh
 
 ENV AUTHORIZED_KEYS **None**
+
+ADD run.sh /run.sh
+RUN chmod +x /*.sh
 
 RUN apt-get update
 RUN apt install -y apt-transport-https

--- a/docker/node/run.sh
+++ b/docker/node/run.sh
@@ -1,0 +1,22 @@
+#!/bin/bash
+
+if [ "${AUTHORIZED_KEYS}" != "**None**" ]; then
+    echo "=> Found authorized keys"
+    mkdir -p /root/.ssh
+    chmod 700 /root/.ssh
+    touch /root/.ssh/authorized_keys
+    chmod 600 /root/.ssh/authorized_keys
+    IFS=$'\n'
+    arr=$(echo ${AUTHORIZED_KEYS} | tr "," "\n")
+    for x in $arr
+    do
+        x=$(echo $x |sed -e 's/^ *//' -e 's/ *$//')
+        cat /root/.ssh/authorized_keys | grep "$x" >/dev/null 2>&1
+        if [ $? -ne 0 ]; then
+            echo "=> Adding public key to /root/.ssh/authorized_keys: $x"
+            echo "$x" >> /root/.ssh/authorized_keys
+        fi
+    done
+fi
+
+exec /usr/sbin/sshd -D

